--- ../installperl	2003-03-26 21:07:13.000000000 +0000
+++ installperl	2003-05-01 11:17:48.000000000 +0000
@@ -3,8 +3,8 @@
 BEGIN {
     require 5.004;
     chdir '..' if !-d 'lib' and -d '..\lib';
-    @INC = 'lib';
-    $ENV{PERL5LIB} = 'lib';
+#    @INC = 'lib';
+#    $ENV{PERL5LIB} = 'lib';
 }
 
 use strict;
@@ -878,7 +878,7 @@
 		print " " if (@opts);
 		print "$file\n";
 	    }
-	    system("strip", @opts, $file);
+	    system("XXSTRIPXX-strip", @opts, $file);
 	} else {
 	    print "# file '$file' skipped\n" if $verbose;
 	}
